File: LeetCode/Python Solutions/Tallest Billboard/tallest.py
from typing import List

class Solution:
    def tallestBillboard(self, rods: List[int]) -> int:
        """
        Finds the maximum possible sum of the shorter two billboards from the given set of rods. 
        The rods list contains the lengths of all the rods.
        Returns the maximum possible sum of the shorter two billboards.
        """
        dp = {0: 0}
        for x in rods:
            dp2 = dict(dp)
            for d, y in dp.items():
                dp2[d + x] = max(dp2.get(d + x, 0), y)
                dp2[abs(d - x)] = max(dp2.get(abs(d - x), 0), y + min(d, x))
            dp = dp2
        return dp[0]

File: LeetCode/Java Solutions/Tallest Billboard/tallest.java
import java.util.*;

class Solution {
    /**
     * Calculates the maximum height possible for a billboard using the given array of rod heights.
     *
     * @param  rods  an array of integers representing the heights of the rods
     * @return       the maximum height possible for the billboard
     */
    public int tallestBillboard(int[] rods) {
        Map<Integer, Integer> dp = new HashMap<>();
        dp.put(0, 0);
        for (int x : rods) {
            Map<Integer, Integer> dp2 = new HashMap<>(dp);
            for (int d : dp.keySet()) {
                dp2.put(d + x, Math.max(dp2.getOrDefault(d + x, 0), dp.get(d)));
                dp2.put(Math.abs(d - x), Math.max(dp2.getOrDefault(Math.abs(d - x), 0), dp.get(d) + Math.min(d, x)));
            }
            dp = dp2;
        }
        return dp.get(0);
    }
}
